File: behavioral/mediator-example.py
class Shop:
    def __init__(self, name):
        self.__name = name
        self.__product = None

    def order(self, product):
        self.__product = product

    def to_str(self):
        return f'name:{self.__name} , product:{self.__product}'


class Delivery:

    @staticmethod
    def collect(merchandises):
        for m in merchandises:
            print(m.to_str())


class Mediator:
    @staticmethod
    def shop_collect(merchandises):
        shops = []
        for m in merchandises:
            s = Shop(m['shop'])
            s.order(m['product'])
            shops.append(s)
        Delivery.collect(shops)


if __name__ == "__main__":
    zara = Shop('Zarah')
    zara.order('that green cool dress')

    brown = Shop('Brown')
    brown.order('cutey shoes')

    Delivery.collect((zara, brown))

    print('- ' * 40)

    Mediator.shop_collect(({'shop': 'Zarah', 'product': 'that green cool dress'},
                           {'shop': 'Brown', 'product': 'cutey shoes'}))

File: behavioral/mediator-example.js
class Shop {
    constructor(name) {
        this.name = name;
        this.product = null;
    }


    order(product) {
        this.product = product;
    }

    toString() {
        return `name:${this.name} , product:${this.product}`;
    }
}

// ---------------------------------------------------------
class Delivery {

    static collect(merchandises) {
        for (let m of merchandises) {
            console.log(m.toString());
        }
    }
}

// ---------------------------------------------------------
class Mediator {
    static shopCollect(merchandises) {
        const shops = [];
        for (let m of merchandises) {
            let s = new Shop(m.shop);
            s.order(m.product);
            shops.push(s);
        }
        Delivery.collect(shops);
    }
}

// ---------------------------------------------------------
const zara = new Shop('Zarah');
zara.order('that green cool dress');

const brown = new Shop('Brown');
brown.order('cutey shoes');

Delivery.collect([zara, brown]);

console.log('- - - - - - - - - - - - - - -');

Mediator.shopCollect([
    { shop: 'Zarah', product: 'that green cool dress' },
    { shop: 'Brown', product: 'cutey shoes' }
]);

